qcodes/instrument_drivers/Lakeshore/lakeshore_base.py

```diff
@@ -0,0 +1,302 @@
+
+    def set_raw(self, value, **kwargs):  # pylint: disable=E0202
+        self.group.set(self, value)
+
+    def get_raw(self, result=None):  # pylint: disable=E0202
+        if not result:
+            self.group.update()
+            return self.raw_value
+        else:
+            return result
+
+
+class Group():
+    def __init__(self, parameters, set_cmd=None, get_cmd=None,
+                 get_parser=None, separator=',', types=None) -> None:
+        self.parameters = OrderedDict((p.name, p) for p in parameters)
+        self.instrument = parameters[0].root_instrument
+        for p in parameters:
+            p.group = self
+        self.set_cmd = set_cmd
+        self.get_cmd = get_cmd
+        if get_parser:
+            self.get_parser = get_parser
+        else:
+            self.get_parser = self._separator_parser(separator, types)
+
+    def _separator_parser(self, separator, types):
+        def parser(ret_str):
+            keys = self.parameters.keys()
+            values = ret_str.split(separator)
+            return dict(zip(keys, values))
+        return parser
+
+    def set(self, set_parameter, value):
+        if any((p.get_latest() is None) for p in self.parameters.values()):
+            self.update()
+        calling_dict = {name: p.raw_value
+                        for name, p in self.parameters.items()}
+        calling_dict[set_parameter.name] = value
+        command_str = self.set_cmd.format(**calling_dict)
+        set_parameter.root_instrument.write(command_str)
+
+    def update(self):
+        ret = self.get_parser(self.instrument.ask(self.get_cmd))
+        # if any(p.get_latest() != ret[] for name, p in self.parameters if p
+        # is not get_parameter):
+        #     log.warn('a value has changed on the device')
+        # TODO(DV): this is odd, but the only way to call the wrapper
+        # accordingly
+        for name, p in list(self.parameters.items()):
+            p.get(result=ret[name])
+
+VAL_MAP_TYPE = ClassVar[Dict[str, int]]
+INVERSE_VAL_MAP_TYPE = ClassVar[Dict[int, str]]
+
+class BaseOutput(InstrumentChannel):
+
+    MODES: ClassVar[Dict[str, int]] = {}
+    RANGES: ClassVar[Dict[str, int]] = {}
+
+
+    def __init__(self, parent, output_name, output_index) -> None:
+        super().__init__(parent, output_name)
+        self.INVERSE_RANGES: Dict[int, str] = {
+            v: k for k, v in self.RANGES.items()}
+        self._has_pid = True
+        self.output_index = output_index
+        self.add_parameter('mode',
+                           val_mapping=self.MODES,
+                           parameter_class=GroupParameter)
+        self.add_parameter('input_channel', get_parser=int,
+                           parameter_class=GroupParameter)
+        self.add_parameter('powerup_enable',
+                           val_mapping={True: 1, False: 0},
+                           parameter_class=GroupParameter)
+        self.output_group = Group([self.mode, self.input_channel,
```

Comment:
I quite like the way the group parameter works. I think we should make this available generally to other instruments as well.